ironfish-cli/src/commands/service/snapshot.test.ts

```diff
@@ -0,0 +1,122 @@
+
+describe('service:snapshot', () => {
+  jest.spyOn(Date, 'now').mockReturnValue(123456789)
+  const mockedFileSize = 10000
+
+  const manifestContent = JSON.stringify(
+    {
+      block_height: 3,
+      checksum: Buffer.from('foo').toString('hex'),
+      file_name: `ironfish_snapshot_${Date.now()}.tar.gz`,
+      file_size: mockedFileSize,
+      timestamp: Date.now(),
+    },
+    undefined,
+    '  ',
+  )
+
+  beforeAll(() => {
+    jest.doMock('@ironfish/sdk', () => {
+      const originalModule = jest.requireActual('@ironfish/sdk')
+
+      const response = {
+        contentStream: jest.fn(async function* () {
+          const stream = [
+            { start: 1, stop: 3 },
+            { start: 1, stop: 3, seq: 3, buffer: Buffer.from('foo') },
+          ]
+
+          for await (const value of stream) {
+            yield value
+          }
+        }),
+      }
+
+      const client = {
+        connect: jest.fn(),
+        snapshotChainStream: jest.fn().mockReturnValue(response),
+      }
+
+      const mockFileSystem = {
+        mkdir: jest.fn(),
+        resolve: jest.fn(),
+        join: jest
+          .fn()
+          .mockReturnValueOnce('testtempdir/blocks')
+          .mockReturnValueOnce('testtempdir/blocks/3')
+          .mockReturnValueOnce(`testtempdir/ironfish_snapshot_${Date.now()}.tar.gz`)
+          .mockReturnValueOnce(`testtempdir/manifest.json`),
+      }
+
+      const module: typeof jest = {
+        ...originalModule,
+        IronfishSdk: {
+          init: jest.fn().mockReturnValue({
+            connectRpc: jest.fn().mockResolvedValue(client),
+            client,
+            fileSystem: mockFileSystem,
+          }),
+          response,
+        },
+      }
+
+      return module
+    })
+
+    jest.mock('aws-sdk', () => {
+      const mockS3 = {
+        upload: jest.fn().mockReturnThis(),
+        promise: jest.fn(),
+      }
+      return { S3: jest.fn(() => mockS3) }
+    })
+
+    jest.mock('tar', () => {
+      return { create: jest.fn() }
+    })
+
+    jest.mock('crypto', () => {
+      const mockHasher = {
+        update: jest.fn().mockReturnThis(),
+        digest: jest.fn().mockReturnValue(Buffer.from('foo')),
+      }
+      return { createHash: jest.fn(() => mockHasher), Hash: jest.fn(() => mockHasher) }
+    })
+
+    jest.mock('fs/promises', () => {
+      const mockFileHandle = {
+        createReadStream: jest.fn().mockReturnValue(['test']),
+      }
+
+      const mockStats = {
+        size: mockedFileSize,
+      }
+
+      return {
+        open: jest.fn().mockReturnValue(mockFileHandle),
+        writeFile: jest.fn(() => Promise.resolve()),
+        mkdtemp: jest.fn().mockReturnValue('testtempdir/'),
+        FileHandle: jest.fn(() => mockFileHandle),
+        stat: jest.fn(() => mockStats),
+      }
+    })
+  })
+
+  afterAll(() => {
+    jest.dontMock('@ironfish/sdk')
+  })
+
+  describe('exports a snapshot of the chain and uploads it', () => {
```

Comment:
Part of the reason you're having trouble writing tests is you aren't thinking of them in a compartmentalized way.

If you have CLI command A that calls function B, C , and D, you generally want to unit test B, C, and D and prove that they fulfill their contract in various conditions. The unit testing allows you to relatively easily vary params and such.

Then you integration test A fully, which tends to be the hardest part because you have to do things like corral output streams and mock file systems.

The most useful and flexible tests tend to be the unit tests. You are focusing in both cases here on integration tests.